DOC update and improve the `sample_weight` entry in the glossary

ogrisel opened the pull request:
As discussed in https://github.com/scikit-learn/scikit-learn/pull/29907#discussion_r1880367391.

- Link to practical usage examples.
- Removed some FIXMEs and TODOs.
- Fixed the description of the `min_samples` in `DBSCAN` and contrasted it with the `min_sample_leaf` / `min_weight_fraction_in_leaf` parameters pair.
- Refine the description of the interplay with the `class_weight` parameter.
- Reference ongoing work and the tracking meta-issue.

ogrisel committed:
DOC update and improve the sample_weight entry in the glossary

ogrisel commented:
For information, for the examples of sample weight usage, we could be more specifice, for instance:

- Survival Analysis: accelerated failure time modeling via reduced penalized linear regression
  - paper: https://www.sciencedirect.com/science/article/pii/S0047259X83710286?via%3Dihub
  - code: https://github.com/sebp/scikit-survival/blob/7c8bdb95fc5df4b24e19342d8021501cba3b0cf9/sksurv/linear_model/aft.py#L184-L185
- Causal Inference: Double Machine Learning
  - paper: https://arxiv.org/abs/1712.04912
  - code: https://github.com/py-why/EconML/blob/d995bf32f7867512d98461b544dba3e66b30a702/econml/dml/dml.py#L175-L181
- Unfairness mitigation: exponentiated gradient reduction
  - paper: http://proceedings.mlr.press/v80/agarwal18a.html
  - code: https://github.com/fairlearn/fairlearn/blob/main/fairlearn/reductions/_exponentiated_gradient/_lagrangian.py#L222

But linking to the papers does not necessarily make it explicit that this can be implemented via scikit-learn's `sample_weight` and the code of those libraries is not always easy to understand without reading the papers first and furthermore I am not sure how to link to a source code snippet while making it easy to check that the link stays relevant over time.

virchan commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,59 @@ See concept :term:`sample property`.
-        In classification, sample weights can also be specified as a function
-        of class with the :term:`class_weight` estimator :term:`parameter`.
+        Weighting samples can be useful in several contexts. For instance, if the
+        training data is not uniformly sampled from the target population, it
+        is possible to correct for this by weighting training data points
+        according to the inverse probability of being sampled for training. It
```
"[...], it can be corrected by weighting the training data points based on the inverse probability of their selection for training."

virchan commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,59 @@ See concept :term:`sample property`.
+        and
+        :ref:`sphx_glr_auto_examples_linear_model_plot_tweedie_regression_insurance_claims.py`).
+
+        Third-party libraries can also use `sample_weight`-capable estimators
```
"Third-party libraries can also use `sample_weight`-compatible estimators"

virchan commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,59 @@ See concept :term:`sample property`.
+        tracks ongoing work to detect and fix remaining discrepancies.
+
+        Furthermore, some estimators have a stochastic fit method. For
+        instance, k-means clustering depends on a random initialization,
```
:class:`cluster.KMeans`

virchan reviewed (commented):
Just wanted to offer a few alternative word choices for consideration.

ogrisel committed:
Improve phrasing

ogrisel commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,60 @@ See concept :term:`sample property`.
+        evaluation.
+
+        At the time of writing, not all scikit-learn estimators correctly
+        implement the  weight-repetition equivalence property. The `#16298 meta
```
```suggestion
        implement the weight-repetition equivalence property. The `#16298 meta
```

ogrisel committed:
Typo

ogrisel commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,60 @@ See concept :term:`sample property`.
+        bagging models randomly resample from the training data, etc. In this
+        case, the sample weight-repetition equivalence property described above
+        does not hold exactly. However, it should hold at least in expectation
+        over the randomness of the fitting procedure.
```
Note to reviewers: I think the equivalence should even hold in distribution rather than just in expectation, but I am not 100% sure this can always be enforced in scikit-learn (yet).

lorentzenchr commented:
Could you wait for my review before merging? Even if it takes 2 weeks?

lorentzenchr reviewed (commented):
Just a high-level review: This PR makes sample weights the longest entry in the glossary. I would rather prefer a concise entry.

ogrisel commented:
> Just a high-level review: This PR makes sample weights the longest entry in the glossary. I would rather prefer a concise entry.

I could move some of the details to a dedicated section of the user guide (maybe part of the estimator API) and keep a short version in the glossary with a link to the user guide for more details if people prefer.

lorentzenchr commented:
FYI: In supervised learning, one of the main reasons for using weights, e.g. in insurance frequency models (Poisson loss), is to account for the different size/time/exposure/... of each observation. The assumptions are usually:
- $Y=\frac{C}{w}$ is the ratio of interest, C is an amount (counts, EUR, kilogram, ...).
- $Var[Y] \sim \frac{1}{w}$

In the end, the weights enters the loss. The idea is that the "correct" weights improve the efficiency of the estimation of the expected loss** (=empirical mean of per observation losses). However, to my knowledge, there is no direct connection between $Var[Y_i]$ and $Var[loss_i]$ , see Section 5.2.2 in https://arxiv.org/abs/2202.12780.

**expected loss = statistical risk

GaelVaroquaux reviewed (commented):
I would also mention reweighting for transfer learning, and define in one sentence what it means.

Also, I find that this paragraph is very useful, but either put it in a fold or in a section of the docs

StefanieSenger commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,60 @@ See concept :term:`sample property`.
         the weight.  Weights may be specified as floats, so that sample weights
```
The second sentence ready rather complicated and can be simplified.

Maybe like this:

`Intuitively, if all weights are integers, using them in a model or scorer is like duplicating each sample as many times as the weight value.`

StefanieSenger commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,60 @@ See concept :term:`sample property`.
+        for model training or a parameter of a :term:`scorer` for model
+        evaluation.
+
+        At the time of writing, not all scikit-learn estimators correctly
```
```suggestion
        At the time of writing (version 1.7), not all scikit-learn estimators correctly
```

StefanieSenger commented on doc/glossary.rst:
```diff
@@ -1854,20 +1854,60 @@ See concept :term:`sample property`.
+        ``class_weight`` are provided, the final weight assigned to a sample is
+        the product of the two.
+
+        `sample_weight` can be both an argument of the estimator's `fit` method
+        for model training or a parameter of a :term:`scorer` for model
+        evaluation.
```
I think this should be one of the first few sentences in this glossary entry.

In accordance with the rest of our terminology, we should talk about these methods _consuming_ sample_weight and mention that there are others that are able to _route_ sample_weight (with a link-reference to metadata routing).

StefanieSenger reviewed (commented):
Thanks @ogrisel. I have added a few comments.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Gael Varoquaux <gael.varoquaux@normalesup.org>
Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

ogrisel committed:
Reorder paragraphs and reformat

ogrisel committed:
Simpler phrasing

ogrisel committed:
Cross-reference metadata-routing

ogrisel commented:
Thanks for the reviews @StefanieSenger @GaelVaroquaux @lorentzenchr. I addressed all the specific feedback. Personally, I am fine with keeping all this info centralized into a largish glossary entry, but I can also split it into a dedicated section in the user guide (where) and cross-reference if people prefer.

betatim reviewed (approved):
I think this is a good entry, even though it is quite long. 

It would be nicer to have a short glossary entry and a dedicated user guide section. But I think we'd need to put in more work to make the user guide section. For example expanding on the third party paragraph from this PR with some links to examples of third-party estimators (I'm wondering if the text as it is, is too brief for people who aren't experts to understand it and too long for those who are experts/third-party implementers).

The final paragraph on KMeans and sampling is another one that I think we could expand in the user guide. At least I am not sure I fully understand what it is trying to tell me without an example.

However, maybe we should merge this now already and make an issue to improve the user-guide and reduce the glossary entry. Mostly because making the user guide entry would be quite a bit of additional work

adrinjalali commented on doc/glossary.rst:
```diff
@@ -1848,26 +1848,72 @@ See concept :term:`sample property`.
+        A relative weight for each sample. Intuitively, if all weights are
+        integers, using them in a model or scorer is like duplicating each
+        sample as many times as the weight value. Weights may be specified as
+        floats, so that sample weights are usually equivalent up to a constant
+        positive scaling factor.
```
equivalent to what? I don't understand this sentence.

ogrisel replied:
Invariance to the scale of weights is actually not guaranteed for all estimators or hyperparameter choices. So let's be more generic.

```suggestion
        floats to express fractional relative importance of data points with respect
        to one another others.
```

ogrisel replied:
I did not want to explicitly state that because we are not yet sure to which extent this is true or not: we do haven't started testing this aspect systematically, but we know that this is not the case for many important ones such as `LogisticRegression` and `Ridge`. Both those estimators have a regularization parameter whose effect depends on the scale of the weights.

adrinjalali commented on doc/glossary.rst:
```diff
@@ -1848,26 +1848,72 @@ See concept :term:`sample property`.
+        fairness-related harms based on a given quantitative definition of
+        fairness.
+
+        Some model hyper-parameters are expressed in terms of a discrete number
+        of samples in a region of the feature space. When fitting with sample
+        weights, a count of samples is often automatically converted to a sum
+        of their weights as is the case  for `min_samples` in
+        :class:`cluster.DBSCAN`, for instance. However, this is not always the
+        case. In particular, the ``min_samples_leaf`` parameter in
+        :class:`ensemble.RandomForestClassifier` does not take weights into
+        account. One should instead pass `min_weight_fraction_leaf` to
+        :class:`ensemble.RandomForestClassifier` to specify the minimum sum of
+        weights of samples in a leaf.
```
I think this information is best placed in the docstrings of those classes, not here, or here keep only the first 3-4 lines

ogrisel replied:
```suggestion
        Some model hyper-parameters are expressed in terms of a discrete number
        of samples in a region of the feature space. When fitting with sample
        weights, a count of samples is often automatically converted to a sum
        of their weights, but this is not always the case. Please refer to the
        docstring for details.
```

adrinjalali commented on doc/glossary.rst:
```diff
@@ -1848,26 +1848,72 @@ See concept :term:`sample property`.
+        and
+        :ref:`sphx_glr_auto_examples_linear_model_plot_tweedie_regression_insurance_claims.py`).
+
+        Third-party libraries can also use `sample_weight`-compatible
+        estimators as building blocks to reduce a specific statistical task
+        into a weighted regression or classification task. For instance sample
+        weights can be constructed to adjust a time-to-event model for
+        censoring in a predictive survival analysis setting. In causal
+        inference, it is possible to reduce a conditional average treatment
+        effect estimation task to a weighted regression task under some
+        assumptions. Sample weights can also be used to mitigate
+        fairness-related harms based on a given quantitative definition of
+        fairness.
```
This paragraph doesn't feel like it belongs here.

ogrisel replied:
Personally, I find this paragraph very important: it grounds an abstract concept into practical application cases. Have those use cases in mind is helpful both for users and contributors.

ogrisel committed:
Apply suggestions from code review

ogrisel commented on doc/glossary.rst:
```diff
@@ -1848,26 +1848,56 @@ See concept :term:`sample property`.
+        ``class_weight`` are provided, the final weight assigned to a sample is
+        the product of the two.
+
+        At the time of writing (version 1.7), not all scikit-learn estimators
```
Since 1.7 was released, this doc will be published on scikit-learn.org/stable no sooner than 1.8.

```suggestion
        At the time of writing (version 1.8), not all scikit-learn estimators
```

ogrisel committed:
Apply suggestions from code review

lucyleeow reviewed (commented):
Thanks for this, it's a great improvement! Not expert here, so can only make readability remarks and ask some questions.

One thought about including the cases where weights could be used is that are we confident our implementation of weights match the use cases? I really do not know the answer to this. In the stata docs on analytical weights they say:

```
Most commands that allow aweights handle them in this manner. That is, if you specify aweights,
they are
1. normalized to sum to N and then
2. inserted in the calculation formulas in the same way as fweights.
```

not sure if we necessarily need to normalize but I think this implies that these weights are scale invariant, which I don't think all our implementations are...?

Also, this may be a can of worms we do not want to open, but do we want to make any general remarks about handling of negative weights? I think we do not dis-allow it. Apparently it is useful in high energy physics https://github.com/scikit-learn/scikit-learn/issues/12464

(Also +0.5 to expanding more in the user guide and trying to cut the glossary section down, but admit this would be a bit of work. Happy for this to go in and iterate later, as it is already a good improvement from current)

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Lucy Liu <jliu176@gmail.com>

ogrisel committed:
Merge branch 'main' into enh-sample-weight-glossary and address some review comments.

ogrisel committed:
Remove sentence about modeling the frequency per unit of time

adrinjalali reviewed (approved):
Thanks @ogrisel . LGTM. 

I'll let another set of eyes have a look and possibly merge.

lucyleeow commented on doc/glossary.rst:
```diff
@@ -1855,25 +1855,51 @@ See concept :term:`sample property`.
+        data point as many times as the weight value. Weights may also be
+        specified as floats: many estimators and scorers are invariant to a
+        rescaling of the weights by a constant positive factor but there are
+        exceptions.
```
Sorry but I don't quite follow how "Weights may also be specified as floats" relates to "many estimators and scorers are invariant to a rescaling of the weights by a constant positive factor but there are exceptions".

ogrisel replied:
Because floating-point valued weights could be used to express relative weights as opposed to absolute (integer) repetitions if you divide integer weights by their sum across the training set, for instance. However, there are known case where this is not equivalent (and we are still not sure if it's a bug or a feature).

lucyleeow replied:
I *think* I see what you are saying.
Integer weights could also be expressed by a float e.g., float 'relative' weights of [0.1, 0.2, 0.3], can be rescaled by a constant factor of 10 and would become integer weights [1, 2, 3] ?

I guess if weights were scale invariant, the above weights would effectively be the same.

I was recently reading this conversation: https://github.com/numpy/numpy/issues/8935#issuecomment-1398689411 (it's about weights specifically in quantiles) but it made me question if some interpretations of weight *is* sample size dependent (i.e. scale variant)..?

Sorry, this is probably not helpful in moving this forward.

ogrisel replied:
Yes, the point is that many scikit-learn estimators have a `fit` method that is naturally sample_weight-scale invariant, but there are exceptions (sometimes only for some combinations of constructor param values). We would need to conduct an exhaustive survey similarly to what we did to check the validity of the integer frequency semantics.

For estimators that are weight-scale invariant, the semantics of fractional weights (passed as floating point values) is easy to define: they are the same as the integer frequency weights because dividing or multiplying by an arbitrary constant that does not impact the results.

For estimators that are not weight scale invariant, the semantics of non-integer weights is not easy to define, but I don't want to make the glossary entry too complex.

lucyleeow commented on doc/glossary.rst:
```diff
@@ -1855,25 +1855,51 @@ See concept :term:`sample property`.
+        rescaling of the weights by a constant positive factor but there are
+        exceptions.
+
+        `sample_weight` can be both an argument of the estimator's `fit` method
```
nitpick

```suggestion
        `sample_weight` can be both an argument of the estimator's :term:`fit` method
```

lucyleeow reviewed (approved):
Sorry for delay, just a question and nitpick, otherwise LGTM.

I will make an issue just to document that we want to add a user guide for sample weights.

ogrisel committed:
Update doc/glossary.rst

Co-authored-by: Lucy Liu <jliu176@gmail.com>

ogrisel committed:
Merge branch 'main' into enh-sample-weight-glossary

ogrisel committed:
Expand a bit on floating point weights

lucyleeow commented on doc/glossary.rst:
```diff
@@ -1855,25 +1855,53 @@ See concept :term:`sample property`.
+        correctly implement the weight-repetition equivalence property. The
+        `#16298 meta issue
+        <https://github.com/scikit-learn/scikit-learn/issues/16298>`_ tracks
+        ongoing work to detect and fix remaining discrepancies.
```
The 'see below' was referring to this part, but I think this is fine as is!

lucyleeow reviewed (approved):
LGTM, will merge tomorrow, just in case @ogrisel you want to do any last changes.

lucyleeow commented:
Thanks for working on this @ogrisel !
